packages/smooth_app/lib/cache/files/file_cache_manager_impl.dart

```diff
@@ -0,0 +1,114 @@
+import 'dart:io';
+import 'dart:typed_data';
+
+import 'package:flutter/foundation.dart';
+import 'package:path/path.dart';
+import 'package:smooth_app/cache/files/files_cache.dart';
+
+class FileCacheManagerImpl extends FileCache {
+  FileCacheManagerImpl({
+    required this.rootDirectory,
+    required this.subFolderName,
+  }) : assert(subFolderName.isNotEmpty);
+
+  final Future<Directory> rootDirectory;
+  final String subFolderName;
+
+  /// This field will be lazily assigned after a call to [initCache]
+  Directory? _directory;
+
+  @protected
+  Future<void> initCache() async {
+    if (_directory == null) {
+      _directory = Directory(
+        join(
+          await rootDirectory.then((Directory d) => d.absolute.path),
+          'files_cache',
+          subFolderName,
+        ),
+      );
+      await _directory!.create(recursive: true);
+    }
+  }
+
+  Future<void> _ensureInitialized() async {
+    if (_directory == null) {
+      await initCache();
+    }
+  }
+
+  Future<File> _getFilePath(String key) async {
+    await _ensureInitialized();
+    return File(join(_directory!.absolute.path, key));
+  }
+
+  @override
+  Future<Uint8List?> get(String key) async {
+    final File file = await _getFilePath(key);
+
+    if (file.existsSync()) {
+      return file.readAsBytes();
+    } else {
+      return null;
+    }
+  }
+
+  @override
+  Future<bool> containsKey(String key) {
+    return _getFilePath(key).then((File file) => file.existsSync());
+  }
+
+  @override
+  Future<bool> put(
+    String key,
+    Uint8List item, {
+    bool overrideExistingItem = true,
+  }) async {
+    if (await containsKey(key) && !overrideExistingItem) {
+      return false;
+    }
+
+    // Will erase existing file
+    final File file = await _getFilePath(key);
+
+    try {
+      await file.writeAsBytes(
+        item,
+        mode: FileMode.writeOnly,
+        flush: true,
+      );
+    } on FileSystemException {
+      return false;
+    }
+
+    return true;
+  }
+
+  @override
+  Future<int> get length => _ensureInitialized().then(
+        (_) => _directory!.list().length,
```

Comment:
I had bad experiences with this kind of syntax: are you sure the `length` will be returned to the `Future<int>` method?

Reply:
It's the same as having two "await":

```
await _ensureInitialized()
return await _directory!.list().length
```

Reply:
I've just tested, OK it's good. Maybe a `then<int>` would make things more obvious.